editor/editor_command_palette.cpp

```diff
@@ -0,0 +1,254 @@
+	if (commands.has(p_command_name)) {
+		print_line("Command: " + p_command_name + " found");
+	} else {
+		print_line(p_command_name + " not found");
+	}
+	commands[p_command_name].callable.call_deferred(nullptr, 0);
+}
+void EditorCommandPalette::register_shortcuts_as_command() {
+	const String *p_key = nullptr;
+	p_key = unregisterd_shortcuts.next(p_key);
+	while (p_key != nullptr) {
+		String command_name = unregisterd_shortcuts[*p_key].first;
+		Ref<Shortcut> p_shortcut = unregisterd_shortcuts[*p_key].second;
+		Ref<InputEventShortcut> ev;
+		ev.instantiate();
+		ev->set_shortcut(p_shortcut);
+		add_command(command_name, *p_key, callable_mp(EditorNode::get_singleton()->get_viewport(), &Viewport::unhandled_input), varray(ev, false));
+		p_key = unregisterd_shortcuts.next(p_key);
+	}
+	unregisterd_shortcuts.clear();
+}
+
+Ref<Shortcut> EditorCommandPalette::add_shortcut_command(const String &p_command, const String &p_key, Ref<Shortcut> p_shortcut) {
+	if (is_inside_tree()) {
+		Ref<InputEventShortcut> ev;
+		ev.instantiate();
+		ev->set_shortcut(p_shortcut);
+		add_command(p_command, p_key, callable_mp(EditorNode::get_singleton()->get_viewport(), &Viewport::unhandled_input), varray(ev, false));
+	} else {
+		const String command_name = String(p_command);
+		const String key_name = String(p_key);
+		Pair p_pair = Pair(command_name, p_shortcut);
+		unregisterd_shortcuts[key_name] = p_pair;
+	}
+	return p_shortcut;
+}
+
+EditorCommandPalette *EditorCommandPalette::get_singleton() {
+	if (singleton == nullptr) {
+		singleton = memnew(EditorCommandPalette);
+	}
+	return singleton;
+}
+
+EditorCommandPalette::EditorCommandPalette() {
+	allow_multi_select = false;
+
+	VBoxContainer *vbc = memnew(VBoxContainer);
+	vbc->connect("focus_exited", callable_mp(this, &EditorCommandPalette::_hide_command_palette));
+	// todo: add theme change signaal handler
+	add_child(vbc);
+
+	command_search_box = memnew(LineEdit);
+	command_search_box->set_placeholder("search for a command");
+	command_search_box->set_placeholder_alpha(0.5);
+	command_search_box->connect("text_changed", callable_mp(this, &EditorCommandPalette::_text_changed));
+	command_search_box->connect("gui_input", callable_mp(this, &EditorCommandPalette::_sbox_input));
+	command_search_box->connect("text_submitted", callable_mp(this, &EditorCommandPalette::_text_confirmed));
+	vbc->add_margin_child(TTR("Search:"), command_search_box);
+	register_text_enter(command_search_box);
+
+	search_options = memnew(Tree);
+	search_options->connect("item_activated", callable_mp(this, &EditorCommandPalette::_confirmed));
+	search_options->create_item();
+	search_options->set_hide_root(true);
+	search_options->set_hide_folding(true);
+	search_options->add_theme_constant_override("draw_guides", 1);
+	vbc->add_margin_child(TTR("Matches:"), search_options, true);
+}
+
+Ref<Shortcut> ED_SHORTCUT_AND_COMMAND(String p_command, const String &p_path, const String &p_name, uint32_t p_keycode) {
```

Comment:
As `p_command` if very often the same as p_name, I would likely move p_command as the last argument and give it a default value of `''`. Then, if it is empty, we use `p_name` instead of `p_command` as the command name.

Also, I would rename it `p_command_name`

Reply:
sure